hibernate-core/src/main/java/org/hibernate/engine/spi/CollectionEntry.java

```diff
@@ -59,89 +63,109 @@ public final class CollectionEntry implements Serializable {
 	private transient CollectionPersister currentPersister;
 	private transient Object currentKey;
 
-	/**
-	 * For newly wrapped collections, or dereferenced collection wrappers
-	 */
-	public CollectionEntry(CollectionPersister persister, PersistentCollection<?> collection) {
+	private CollectionEntry() {
```

Comment:
Changes to constructors should be reverted. In this particular case, I'd prefer if you inline the code of the methods `setLoadedPerPersister` and `setRole` into the constructors, even if that "duplicates" some code. It's IMO not a big deal, as it's just simple null checks.